File: data_structure/binaryTree.c
#include <stdio.h>
#include <stdlib.h>

struct node
{
    int data;
    struct node* left;
    struct node* right;
};

int max(int a, int b);
void insert(struct node** root, int data);
void preOrder(struct node* root);
void inOrder(struct node* root);
void postOrder(struct node* rood);
int size(struct node* root);
int height(struct node* root);
int sumOfWeight(struct node* root);
int maxPathWeight(struct node* root);
void mirror(struct node** root);
void destruct(struct node** root);

int main() {
    int numTestCases;
    scanf("%d", &numTestCases);
    while (numTestCases--) {
        int num, i;
        struct node* root = NULL;

        scanf("%d", &num);
        for (i=0; i<num; i++){
            int data;
            scanf("%d", &data);
            insert(&root, data);
        }
        // printf("%d\n", size(root));
        // printf("%d\n", height(root));
        // printf("%d\n", sumOfWeight(root));
        // printf("%d\n", maxPathWeight(root));
        // mirror(&root);
        preOrder(root); printf("\n");
        inOrder(root); printf("\n");
        postOrder(root); printf("\n");
        destruct(&root);
        if (root == NULL)
            printf("0\n");
        else
            printf("1\n");

    }
    return 0;
}

int max(int a, int b){
    return (a>b)? a: b;
}

// 데이터삽입
void insert(struct node** root, int data) {
    if (*root == NULL) {
        *root = (struct node*)malloc(sizeof(struct node));
        (*root)->data = data;
        (*root)->left = NULL;
        (*root)->right = NULL;
    }
    else if (data < (*root)->data){
        insert(&((*root)->left), data);
    }
    else if (data > (*root)->data){
        insert(&((*root)->right), data);
    }
}

// 전위탐색
void preOrder(struct node* root) {
    if (root == NULL) return;
    else {
        printf("%d ", root->data);
        preOrder(root->left);
        preOrder(root->right);
    }
}

// 중위탐색
void inOrder(struct node* root) {
    if (root == NULL) return;
    else {
        inOrder(root->left);
        printf("%d ", root->data);
        inOrder(root->right);
    }
}

// 후위탐색
void postOrder(struct node* root) {
    if (root == NULL) return;
    else {
        postOrder(root->left);
        postOrder(root->right);
        printf("%d ", root->data);
    }
}

// 노드의 개수
// int size(struct node* root) {

// }

// // 높이 (height)
// int height(struct node* root) {
    
// }

// // 미러 이지미로 변환
// void mirror(struct node** root) {
    
// }

// // 노드에 저장된 데이터 값의 합
// int sumOfWeight(struct node* root) {
    
// }

// // 루트노드부터 단말노드까지의 경로 상의 데이터 최대합
// int maxPathWeight(struct node* root) {
    
// }

// 트리노드의 동적 메모리 해제
void destruct(struct node** root) {
    if(*root == NULL) return;
    destruct(&((*root)->left));
    destruct(&((*root)->right));
    free(*root);
    *root = NULL;
}

File: data_structure/binaryTree.py
class Node:
    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None

class BinaryTree:
    def __init__(self, value=None):
        self.root = Node(value)

    def insert(self, value):
        if self.root.value is None:
            self.root.value = value
            return
        else:
            self.insert_recursive(self.root, value)

    def insert_recursive(self, node, value):
        if value < node.value:
            if node.left is None:
                node.left = Node(value)
            else:
                self.insert_recursive(node.left, value)
        else:
            if node.right is None:
                node.right = Node(value)
            else:
                self.insert_recursive(node.right, value)

    def size(self, node):
        if node is None: return 0
        return 1 + self.size(node.left) + self.size(node.right)

    def height(self, node):
        if node is None: return -1
        return 1 + max(self.height(node.left), self.height(node.right))

    def sumOfWeight(self, node):
        if node is None: return 0
        return node.value + self.sumOfWeight(node.left) + self.sumOfWeight(node.right)

    def maxPathWeight(self, node):
        if node is None: return 0
        return node.value + max(self.maxPathWeight(node.left), self.maxPathWeight(node.right))

    def mirror(self, node):
        if node is None: return
        tmp = node.left
        node.left = node.right
        node.right = tmp
        self.mirror(node.left)
        self.mirror(node.right)

    def preOrder(self, node):
        if node:
            print(node.value, end=' ')
            self.preOrder(node.left)
            self.preOrder(node.right)

    def inOrder(self, node):
        if node:
            self.inOrder(node.left)
            print(node.value, end=' ')
            self.inOrder(node.right)

    def postOrder(self, node):
        if node:
            self.postOrder(node.left)
            self.postOrder(node.right)
            print(node.value, end=' ')

    def destruct(self):
        self.destruct_recursive(self.root)
        self.root = None

    def destruct_recursive(self, node):
        if node is None: return
        self.destruct_recursive(node.left)
        self.destruct_recursive(node.right)
        del node


if __name__ == '__main__':
    # 출력순서: size, height, sumOfWeight, maxPathWeight, (mirror), pre, in, post, destruct
    numTestCases = int(input())

    for _ in range(numTestCases):
        tree = BinaryTree()

        a = list(map(int, input().split()))
        num = a.pop(0)
        for i in range(num):
            tree.insert(a[i])

        print(tree.size(tree.root))
        print(tree.height(tree.root))
        print(tree.sumOfWeight(tree.root))
        print(tree.maxPathWeight(tree.root))
        tree.mirror(tree.root)
        tree.preOrder(tree.root)
        print()
        tree.inOrder(tree.root)
        print()
        tree.postOrder(tree.root)
        tree.destruct()
        print()
        if tree.root is None: print(0)
        else: print(1)